File: src/plan-store.ts
// ---------------------------------------------------------------------------
// The plan document: what a stored project set looks like, and the wire it
// travels on.
//
// `storage.ts` orchestrates (load / save / fall back / mirror); this file holds
// the shape everything agrees on and the raw WebSocket calls under it. The
// rules the wire enforces are worth reading on their own:
//
//  * A read that FAILED is never reported as "empty". An empty answer is a
//    licence to save on top, and saving on top of a failed read is how one
//    device wipes every project on every other device.
//  * A write carries the version the editor started from (`base_version`). The
//    integration refuses the write when the document has moved on since, and
//    hands back the document that won — so two tablets saving a second apart no
//    longer means the second one silently deletes the first one's work.
//  * A refusal says WHY (no rights / conflict / too large / store unavailable),
//    in Russian, so the interface can say something true instead of "не смог".
// ---------------------------------------------------------------------------

import type { FloorPlan, HomeAssistant } from './types';

export interface StoredProjects {
  active?: string;
  projects: Record<string, FloorPlan>;
  /** Hashed PIN that gates Edit mode (casual tamper-protection on a kiosk).
   *  Not cryptographically strong — just stops a passer-by from editing.
   *  Stored in its OWN document by the integration; filled in here on load and
   *  taken back out before the plan is written. */
  editPin?: string;
  /** Version of the install-wide document this set was read from, stamped by
   *  the integration. Sent back on save as `base_version` so a save cannot
   *  overwrite work done on another device in the meantime. Absent means the
   *  document came from somewhere that does not version (per-user storage,
   *  localStorage, or an integration older than this). */
  version?: number;
}

/** Small non-cryptographic hash so the edit PIN isn't stored in plain text.
 *  (Casual protection only — clearing browser/HA storage resets it.) */
export function hashPin(s: string): string {
  let h = 5381;
  for (let i = 0; i < s.length; i++) h = ((h << 5) + h + s.charCodeAt(i)) >>> 0;
  return 'h' + h.toString(36);
}

export interface ProjectInfo {
  id: string;
  name: string;
}

export function valid(plan: any): plan is FloorPlan {
  return !!plan && Array.isArray(plan.floors) && plan.floors.length > 0;
}

export function listProjects(data: StoredProjects): ProjectInfo[] {
  return Object.entries(data.projects)
    .filter(([, p]) => valid(p))
    .map(([id, p]) => ({ id, name: p.name || id }))
    .sort((a, b) => a.name.localeCompare(b.name));
}

export function newProjectId(): string {
  try {
    if (typeof crypto !== 'undefined' && (crypto as any).randomUUID) {
      return 'p' + (crypto as any).randomUUID().replace(/-/g, '').slice(0, 12);
    }
  } catch {
    /* fall through */
  }
  return `p${Date.now().toString(36)}${Math.floor(Math.random() * 1e9).toString(36)}`;
}

/** A fresh empty plan to draw on from scratch. */
export function blankPlan(name = 'New Plan'): FloorPlan {
  return {
    name,
    wallHeight: 2.6,
    floors: [{ name: 'Ground Floor', elevation: 0, wallHeight: 2.6, walls: [], rooms: [], furniture: [], bindings: [] }],
  };
}

// --- The wire --------------------------------------------------------------

export const PLAN_WS_GET = 'bms_floorplan/plan/get';
export const PLAN_WS_SET = 'bms_floorplan/plan/set';
/** The edit PIN has a document of its own — a plan save must not touch it. */
export const PIN_WS_GET = 'bms_floorplan/pin/get';
export const PIN_WS_SET = 'bms_floorplan/pin/set';

export function errMsg(e: any): string {
  return String(e?.message ?? e?.code ?? e ?? 'unknown error');
}

/** True when the failure means "this endpoint does not exist here" (an older or
 *  absent integration) rather than "the read went wrong". The difference
 *  matters: absent is a normal state we fall back from, a failed read is NOT —
 *  treating it as "empty" is how a save wipes every other device's projects. */
export function isUnavailable(e: any): boolean {
  const code = String(e?.code ?? '').toLowerCase();
  if (code === 'unknown_command' || code === 'not_found' || code === 'integration_unloaded') {
    return true;
  }
  const msg = String(e?.message ?? e ?? '').toLowerCase();
  return msg.includes('unknown command') || msg.includes('unknown_command') || msg.includes('not found');
}

export type ReadStatus =
  | 'ok' // read succeeded and there is data
  | 'empty' // read succeeded, nothing stored yet
  | 'unavailable' // the endpoint does not exist here
  | 'error'; // the read FAILED — the data may well exist

export interface ReadOutcome {
  status: ReadStatus;
  data?: StoredProjects;
  error?: string;
}

/** Why a write did not happen. Kept coarse on purpose: each value maps to one
 *  sentence a human can act on. */
export type SaveFailReason =
  | 'forbidden' // not an administrator
  | 'conflict' // someone saved first
  | 'too_large' // over the integration's ceiling
  | 'invalid' // the document was refused as malformed
  | 'unavailable' // the store could not be reached at all
  | 'failed'; // anything else — the raw message is carried alongside

const REASON_TEXT: Record<SaveFailReason, string> = {
  forbidden: 'Нет прав: изменять общий план может только администратор.',
  conflict: 'План уже изменён на другом устройстве — запись отменена, чтобы не стереть чужие правки.',
  too_large: 'План слишком большой для хранилища — уменьшите фотографии.',
  invalid: 'Хранилище не приняло документ плана.',
  unavailable: 'Хранилище плана недоступно.',
  failed: 'Не удалось сохранить план.',
};

function reasonFor(e: any): SaveFailReason {
  const code = String(e?.code ?? '').toLowerCase();
  if (code === 'unauthorized' || code === 'not_admin') return 'forbidden';
  if (code === 'too_large') return 'too_large';
  if (code === 'storage_unavailable') return 'unavailable';
  if (code === 'version_conflict') return 'conflict';
  if (
    code === 'invalid_plan' ||
    code === 'not_json' ||
    code === 'invalid_pin' ||
    code === 'invalid_base_version' ||
    code === 'invalid_format'
  ) {
    return 'invalid';
  }
  const msg = String(e?.message ?? e ?? '').toLowerCase();
  if (msg.includes('unauthorized') || msg.includes('администратор')) return 'forbidden';
  return 'failed';
}

/** Home Assistant's own refusals are English; ours are already Russian. */
function humanError(reason: SaveFailReason, serverMessage?: string): string {
  if (reason === 'forbidden') return REASON_TEXT.forbidden;
  const msg = (serverMessage ?? '').trim();
  return msg || REASON_TEXT[reason];
}

/** Accept both the bare `{active, projects}` document and a `{found, data}`
 *  envelope, so the same reader works for plan/get and legacy/get. */
export function asSet(raw: any): StoredProjects | null {
  if (!raw || typeof raw !== 'object') return null;
  if (raw.projects && typeof raw.projects === 'object') return raw as StoredProjects;
  const inner = (raw as any).data;
  if (inner && typeof inner === 'object' && inner.projects) return inner as StoredProjects;
  return null;
}

/** The install-wide copy served by our integration. */
export async function readShared(hass: HomeAssistant, cmd = PLAN_WS_GET): Promise<ReadOutcome> {
  if (!hass.callWS) return { status: 'unavailable' };
  try {
    const raw: any = await hass.callWS({ type: cmd });
    const data = asSet(raw);
    return data ? { status: 'ok', data } : { status: 'empty' };
  } catch (e) {
    return isUnavailable(e) ? { status: 'unavailable' } : { status: 'error', error: errMsg(e) };
  }
}

/** A `frontend/get_user_data` document (ours, or — read-only — the old one). */
export async function readUserData(hass: HomeAssistant, key: string): Promise<ReadOutcome> {
  if (!hass.callWS) return { status: 'unavailable' };
  try {
    const res: any = await hass.callWS({ type: 'frontend/get_user_data', key });
    const data = asSet(res?.value);
    return data ? { status: 'ok', data } : { status: 'empty' };
  } catch (e) {
    return isUnavailable(e) ? { status: 'unavailable' } : { status: 'error', error: errMsg(e) };
  }
}

export interface WriteOutcome {
  ok: boolean;
  /** The endpoint isn't there (older/absent integration) — not a failure of ours. */
  unavailable?: boolean;
  /** Someone else saved first. NOTHING was written; our copy is the stale one. */
  conflict?: boolean;
  reason?: SaveFailReason;
  /** Ready-to-show Russian text. */
  error?: string;
  /** Raw code from the integration, for logs. */
  code?: string;
  /** Version now stored — after a successful write, or the winner's on a conflict. */
  version?: number;
  /** On a conflict: the document that is actually stored, so the interface can
   *  offer to reload instead of asking the operator to guess. */
  current?: StoredProjects;
}

function failure(e: any): WriteOutcome {
  if (isUnavailable(e)) return { ok: false, unavailable: true, error: errMsg(e), reason: 'unavailable' };
  const reason = reasonFor(e);
  return { ok: false, reason, error: humanError(reason, errMsg(e)), code: String(e?.code ?? '') };
}

/** Mirror the set to the install-wide copy.
 *
 *  `baseVersion` is the version this edit started from. Leaving it undefined is
 *  the blind write the card did before versioning existed — used only when the
 *  document we loaded carried no version at all. */
export async function writeShared(
  hass: HomeAssistant,
  data: StoredProjects,
  baseVersion?: number,
): Promise<WriteOutcome> {
  if (!hass.callWS) return { ok: false, unavailable: true, reason: 'unavailable' };
  const msg: Record<string, unknown> = { type: PLAN_WS_SET, value: data };
  if (typeof baseVersion === 'number') msg.base_version = baseVersion;
  try {
    const res: any = await hass.callWS(msg);
    if (res && res.ok === false) {
      // A refusal that carries data comes back as a RESULT, because a WebSocket
      // error can only carry a code and a message — and on a conflict the
      // caller needs the document that won. In a result the code travels as
      // `error`, so it is renamed here before classifying: reading the wrong
      // field turns a conflict into a nameless failure.
      const reason = reasonFor({ code: res.error, message: res.message });
      return {
        ok: false,
        conflict: reason === 'conflict',
        reason,
        error: humanError(reason, res.message),
        code: String(res.error ?? ''),
        version: typeof res.version === 'number' ? res.version : undefined,
        current: asSet(res.current) ?? undefined,
      };
    }
    return { ok: true, version: typeof res?.version === 'number' ? res.version : undefined };
  } catch (e) {
    // An integration older than base_version rejects the unknown field outright.
    // Retry once without it rather than refusing to save at all.
    if (typeof baseVersion === 'number' && reasonFor(e) === 'invalid') {
      return writeShared(hass, data);
    }
    return failure(e);
  }
}

export interface PinOutcome {
  status: ReadStatus;
  /** The stored hash, or null when Edit mode is not locked. */
  pin?: string | null;
  error?: string;
}

/** Read the PIN from its own document. Open to every signed-in user: the card
 *  has to know whether Edit mode is locked before it can ask for the PIN. */
export async function readPin(hass: HomeAssistant): Promise<PinOutcome> {
  if (!hass.callWS) return { status: 'unavailable' };
  try {
    const res: any = await hass.callWS({ type: PIN_WS_GET });
    const pin = typeof res?.pin === 'string' && res.pin ? res.pin : null;
    return { status: 'ok', pin };
  } catch (e) {
    return isUnavailable(e) ? { status: 'unavailable' } : { status: 'error', error: errMsg(e) };
  }
}

/** Set (or, with null, clear) the PIN. Admin-only in the integration. */
export async function writePin(hass: HomeAssistant, pin: string | null): Promise<WriteOutcome> {
  if (!hass.callWS) return { ok: false, unavailable: true, reason: 'unavailable' };
  try {
    await hass.callWS({ type: PIN_WS_SET, value: pin });
    return { ok: true };
  } catch (e) {
    return failure(e);
  }
}

File: custom_components/bms_floorplan/plan_store.py
"""The two documents this integration owns, and the rules for writing them.

``plan_api`` is the API surface (WebSocket commands and the HTTP view); this is
the storage layer underneath it, kept apart so neither file grows past the point
where it can be read in one sitting.

Two documents live here:

* THE PLAN — one per install, versioned. Every write stamps a monotonic
  ``version`` and an ``updated_at``. A writer says which version it started
  from; if the document has moved on since, the write is refused and the caller
  is handed the document that won. Before this, two tablets saving a second
  apart meant the second one silently deleted the first one's projects.
* THE EDIT PIN — its own document. It used to sit inside the plan, so every plan
  save rewrote it and a device whose copy predated the PIN erased it.

A read that FAILED raises rather than returning "empty". That distinction is the
guard the whole package rests on: an empty answer is a licence to save on top,
and saving on top of a failed read is how one device wipes the install.
"""

from __future__ import annotations

import logging
from datetime import datetime, timezone
from typing import Any

from homeassistant.core import HomeAssistant
from homeassistant.helpers.storage import Store

from .const import (
    DATA_PIN_STORE,
    DOC_UPDATED_KEY,
    DOC_VERSION_KEY,
    DOMAIN,
    LEGACY_PLAN_PIN_KEY,
    PIN_STORAGE_KEY,
    PIN_STORAGE_VERSION,
    STORAGE_KEY,
    STORAGE_VERSION,
)
from .validation import PlanValidationError, validate_pin, validate_plan_document

_LOGGER = logging.getLogger(__name__)

# Failure codes. They travel to the browser, so the card can say WHY a save did
# not happen instead of the single "не смог" it used to show.
ERR_CONFLICT = "version_conflict"
ERR_UNAVAILABLE = "storage_unavailable"

CONFLICT_MSG = (
    "План уже изменён на другом устройстве. Ваша запись отменена, "
    "чтобы не стереть чужие правки — обновите план и повторите."
)
UNAVAILABLE_MSG = "Хранилище плана недоступно, повторите позже."
PIN_UNAVAILABLE_MSG = "Хранилище PIN недоступно, повторите позже."


def _now() -> str:
    """Timestamp for ``updated_at`` — UTC, so devices in one house agree."""
    return datetime.now(timezone.utc).isoformat(timespec="seconds")


def doc_version(doc: Any) -> int:
    """Version of a stored document.

    Anything written before versioning existed (or anything malformed) is
    version 0, which is what a first-time writer sends as its base — so old data
    keeps working with no manual migration.
    """
    if not isinstance(doc, dict):
        return 0
    value = doc.get(DOC_VERSION_KEY)
    if isinstance(value, bool) or not isinstance(value, int) or value < 0:
        return 0
    return value


class PlanConflictError(Exception):
    """Someone else saved first. The write did NOT happen."""

    def __init__(self, current: dict, version: int) -> None:
        super().__init__(CONFLICT_MSG)
        self.code = ERR_CONFLICT
        self.message = CONFLICT_MSG
        self.current = current
        self.version = version


class StoreUnavailableError(Exception):
    """The store could not be read or written. NOT the same as 'it is empty'."""

    def __init__(self, message: str = UNAVAILABLE_MSG) -> None:
        super().__init__(message)
        self.code = ERR_UNAVAILABLE
        self.message = message


class PlanStore:
    """Load/save the shared plan document, cached in memory."""

    def __init__(self, hass: HomeAssistant) -> None:
        self._store: Store = Store(hass, STORAGE_VERSION, STORAGE_KEY)
        self._data: dict | None = None
        self._loaded = False

    async def async_load(self) -> dict | None:
        """The stored document, or None when nothing has ever been saved.

        Raises StoreUnavailableError when the read FAILED — see the module
        docstring for why that is not the same answer as an empty plan.
        """
        if self._loaded:
            return self._data
        try:
            data = await self._store.async_load()
        except Exception as err:  # noqa: BLE001 - reported, not swallowed
            _LOGGER.warning("Не удалось прочитать общий план: %s", err)
            # Deliberately NOT cached: a transient failure must not poison every
            # later read for the lifetime of the process.
            raise StoreUnavailableError() from err
        self._data = data if isinstance(data, dict) else None
        self._loaded = True
        return self._data

    async def async_version(self) -> int:
        """Version currently on disk (0 when nothing is stored)."""
        return doc_version(await self.async_load())

    async def async_save(self, data: Any, base_version: int | None = None) -> int:
        """Validate, stamp with the next version, persist. Returns that version.

        ``base_version=None`` is the OLD-CLIENT PATH: a card bundle from before
        this package sends nothing and writes blindly, exactly as it always did.
        Every current client sends it and is protected.
        """
        if not isinstance(data, dict):
            raise PlanValidationError("Ожидался объект плана.")

        current = await self.async_load()
        version = doc_version(current)
        if base_version is not None and base_version != version:
            raise PlanConflictError(current or {}, version)

        doc = dict(data)
        doc[DOC_VERSION_KEY] = version + 1
        doc[DOC_UPDATED_KEY] = _now()
        validate_plan_document(doc)

        await self._store.async_save(doc)
        # Cache only after the write actually landed, so a failed save cannot
        # leave this process serving a document that is not on disk.
        self._data = doc
        self._loaded = True
        return version + 1


class PinStore:
    """The edit PIN, in a document of its own.

    It used to sit inside the plan document, which meant every plan save rewrote
    it: a device whose copy of the plan predated the PIN erased the PIN, and an
    admin editing the plan could not avoid touching it.

    What is stored is the hash the card computes in the browser, never the
    digits. Reading is open to any authenticated user because the card has to
    know whether Edit mode is locked before it can ask for the PIN; that hash is
    weak by design (see ``hashPin`` in src/storage.ts) — casual
    tamper-protection for a wall tablet, not a secret.
    """

    def __init__(self, hass: HomeAssistant) -> None:
        self._store: Store = Store(hass, PIN_STORAGE_VERSION, PIN_STORAGE_KEY)
        self._data: dict | None = None
        self._loaded = False

    async def async_load(self) -> dict:
        if self._loaded and self._data is not None:
            return self._data
        try:
            data = await self._store.async_load()
        except Exception as err:  # noqa: BLE001
            _LOGGER.warning("Не удалось прочитать PIN редактора: %s", err)
            raise StoreUnavailableError(PIN_UNAVAILABLE_MSG) from err
        self._data = data if isinstance(data, dict) else {}
        self._loaded = True
        return self._data

    async def async_get(self) -> str | None:
        """The stored PIN hash, or None when Edit mode is not locked."""
        pin = (await self.async_load()).get("pin")
        return pin if isinstance(pin, str) and pin else None

    async def _async_write(self, pin: str | None, adopted: bool = False) -> dict:
        current = await self.async_load()
        doc = {
            "pin": pin,
            DOC_VERSION_KEY: doc_version(current) + 1,
            DOC_UPDATED_KEY: _now(),
            # Latched once the PIN that used to live in the plan has been taken
            # over, so removing the PIN later cannot resurrect the old one from
            # the copy still sitting in the plan document.
            "adopted_from_plan": bool(current.get("adopted_from_plan") or adopted),
        }
        try:
            await self._store.async_save(doc)
        except Exception as err:  # noqa: BLE001
            _LOGGER.error("Не удалось сохранить PIN редактора: %s", err)
            raise StoreUnavailableError(PIN_UNAVAILABLE_MSG) from err
        self._data = doc
        self._loaded = True
        return doc

    async def async_set(self, pin: Any) -> dict:
        """Set or clear the PIN (None / "" clears it)."""
        return await self._async_write(validate_pin(pin))

    async def async_adopt(self, plan: dict | None) -> None:
        """One-time carry-over of a PIN saved before this store existed.

        Copies, does not cut: rewriting the plan document during a READ would
        bump its version under other devices' feet for no reason. The leftover
        key inside the plan is harmless — nothing reads it any more, and the next
        save from a current card drops it.
        """
        if not isinstance(plan, dict):
            return
        legacy = plan.get(LEGACY_PLAN_PIN_KEY)
        if not isinstance(legacy, str) or not legacy.strip():
            return
        try:
            current = await self.async_load()
            if current.get("adopted_from_plan") or current.get("pin"):
                return
            await self._async_write(validate_pin(legacy), adopted=True)
        except (StoreUnavailableError, PlanValidationError) as err:
            # Never break a plan read over the PIN: the next read tries again.
            _LOGGER.warning("Не удалось перенести PIN из плана: %s", err)
            return
        _LOGGER.info("PIN редактора перенесён из документа плана в своё хранилище.")


def pin_store(hass: HomeAssistant) -> PinStore:
    """The install's PIN store, created on first use.

    Kept in hass.data next to the plan store so a reload reuses the same cache.
    """
    data = hass.data.setdefault(DOMAIN, {})
    store = data.get(DATA_PIN_STORE)
    if store is None:
        store = PinStore(hass)
        data[DATA_PIN_STORE] = store
    return store
